genologics/lims.py

```diff
@@ -379,21 +380,24 @@ def _get_instances(self, klass, params=dict()):
             root = self.get(node.attrib['uri'], params=params)
         return result
 
-    def get_batch(self, instances):
-        "Get the content of a set of instances using the efficient batch call."
+    def get_batch(self, instances, force=False):
+        """Get the content of a set of instances using the efficient batch call."""
         if not instances:
             return []
-        klass = instances[0].__class__
+
         root = ElementTree.Element(nsmap('ri:links'))
+        klass = None
         result = []
         needs_request=False
         for instance in instances:
-            try:
-                result.append(self.cache[instance.uri])
-            except:
-                needs_request=True
+            if not klass:
```

Comment:
For your code:
4.496076
32.065622

However with only 1 iteration in the xrange(0, N) the batch wins by a simlar factor:
4.2814475
1.7301395

So I think you've found a bug somewhere related to the repeated batch calls. Somehow it's not detecting that the artifacts are already cached in iterations 1..19.